{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0ec21e80",
   "metadata": {},
   "source": [
    "## Ageing transcriptome meta-analysis reveals similarities and differences between key mammalian tissues, 2021, Aging, Palmer D. et al.<hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2c40fbd",
   "metadata": {},
   "source": [
    "### Introduction\n",
    "This notebook and project by Skoltech students tries to reproduce the results of the original paper studying transcriptomic changes associated with ageing in process.<br>For the sake of time we have focussed on mice only, while original paper studied and compared gene expression association with age for mice, rats and humans. For the analysis the ageing transcriptome datasets were used. These usually represent expreriments where several animals/individuals of different age are studied by RNA-sequencing techniques such as total RNA-seq or microarray studies.<br>For our analysis we used 27 datasets in total majority of which were microarray studies (used datasets as well as metafile can be found on our <a href=\"https://github.com/d-kozhevnikova/Ageing-transcriptome-meta-analysis\">GitHub</a> reporsitory). The datasets were already provided by authors as normalized read counts for each gene (RPKM counts that are log2 transformed). We analyzed three tissues:\n",
    "<ul>\n",
    "<li>Heart tissue (5 datasets)</li>\n",
    "<li>Brain tissue (12 datasets)</li>\n",
    "<li>Muscle tissue (10 datasets)</li>\n",
    "</ul>\n",
    "In general the pipeline for the gene expression analysis looks as following:\n",
    "<ol>\n",
    "<li>Running linear regression analysis for each gene inside each dataset (age of samples is regressed against gene expression)</li>\n",
    "<li>Selecting the genes that are:</li>\n",
    "    <ul>\n",
    "    <li>Statistically significantly assoicated with age (regression line slope)</li>\n",
    "    <li>Are associated with age across multiple datasets (at least two)</li>\n",
    "    </ul>\n",
    "<li>Running meta-regression analysis:</li>\n",
    "    <ul>\n",
    "    <li>Using binomial cimulative test which requires:</li>\n",
    "        <ol>\n",
    "            <li>Prior calculation of probabilities for a random gene to be overexpressed and underexpressed - as average across multiple datastes of a percentage of differentially expressed genes from total number of studied genes</li>\n",
    "            <li>Calculation of binomial probability where <i>N</i> is total number of datasets where gene is studies and <i>K</i> is the number of datasets where gene is statistically significantly over/underexpressed with age</li>\n",
    "        </ol>\n",
    "    <li>Using <a href=\"https://pymare.readthedocs.io/en/latest/generated/pymare.core.meta_regression.html#pymare.core.meta_regression\" >PyMare</a> meta-regression function</li>\n",
    "    </ul>    \n",
    "</ol>\n",
    "Based on this we obtained lists of genes that are differentially (over- and under-) expressed with age across multiple datasets for heart, brain, and muscle tissues individually as well as for combined analysis all together.\n",
    "Further steps of the analysis involved:\n",
    "<ul>\n",
    "<li>Running the EnrichGO analysis for gene lists using <a href=\"https://gseapy.readthedocs.io/en/latest/introduction.html\">gseapy</a> python package</li>\n",
    "<li>Intersecting gene lists between different tissues</li>\n",
    "<li>Intersecting gene lists with <a href=\"https://genomics.senescence.info/genes/index.html\">GenAge</a> database</li> \n",
    "</ul>\n",
    "<br>\n",
    "Below is the code that has been used for the analysis of the datasets. All the packages required for the analysis are mentioned in the first cell of the code.<br>\n",
    "In case of any questions please feel free to reach us out to us:\n",
    "<ul>\n",
    "<li>Telegram: @stappyman</li>\n",
    "<li>E-mail: stapranad@mail.ru</li>\n",
    "</ul><hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41181c7a",
   "metadata": {},
   "source": [
    "### Main code:<br>Preparing the data for the analysis\n",
    "The cell below uploads packages required for data analysis and visualisation. During the analysis we have created a separate conda environment which made the analysis and package uploading easier and quicker. The needed packages can be download with following commands:\n",
    "```Python\n",
    "!pip install \"package_name\"\n",
    "!conda install -c bioconda \"package_name\"\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccb9a8b7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy\n",
    "import statsmodels.api as sm\n",
    "import mpmath\n",
    "import wrapt\n",
    "import pymare\n",
    "from pymare import core\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import statsmodels\n",
    "import mygene\n",
    "import gseapy\n",
    "from gseapy import barplot, dotplot\n",
    "import openpyxl\n",
    "from sklearn import preprocessing\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from rpy2 import robjects\n",
    "from matplotlib_venn import venn3, venn2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f557aac5",
   "metadata": {},
   "source": [
    "### Selecting the datasets that are meeting our criteria:\n",
    "The cell below allows selection of the datasets based on <i>metadata.xlsx</i> file. Datasets can be searched based on the following criteria:\n",
    "<ul>\n",
    "<li>Species (either human, mouse or rat)</li>\n",
    "<li>Experimental technique (either RNAseq, Microarray or all)</li>\n",
    "<li>Tissue (requires a list of tissues)</li>    \n",
    "</ul>\n",
    "It return the list of dataset names we need to use for the analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7a4568d9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AgeMap.Brain.RegressionInput.txt',\n",
       " 'AgeMap.Heart.RegressionInput.txt',\n",
       " 'AgeMap.Muscle.RegressionInput.txt',\n",
       " 'GDS1311.Brain.RegressionInput.txt',\n",
       " 'GDS2082.Brain.RegressionInput.txt',\n",
       " 'GDS2612.Muscle.RegressionInput.txt',\n",
       " 'GDS2912.Brain.RegressionInput.txt',\n",
       " 'GDS2972.Heart.RegressionInput.txt',\n",
       " 'GDS2973.Brain.RegressionInput.txt',\n",
       " 'GDS2996.Muscle.RegressionInput.txt',\n",
       " 'GDS3620.Brain.RegressionInput.txt',\n",
       " 'GDS40.Heart.RegressionInput.txt',\n",
       " 'GDS4804.Muscle.RegressionInput.txt',\n",
       " 'GDS4892.Muscle.RegressionInput.txt',\n",
       " 'GDS4904.Muscle.RegressionInput.txt',\n",
       " 'GSE11291.Brain.RegressionInput.txt',\n",
       " 'GSE11291.Heart.RegressionInput.txt',\n",
       " 'GSE11291.Muscle.RegressionInput.txt',\n",
       " 'GSE12480.Heart.RegressionInput.txt',\n",
       " 'GSE19676.Muscle.RegressionInput.txt',\n",
       " 'GSE34378.Brain.RegressionInput.txt',\n",
       " 'GSE48911.Brain.RegressionInput.txt',\n",
       " 'GSE55163.Muscle.RegressionInput.txt',\n",
       " 'GSE61915.Brain.RegressionInput.txt',\n",
       " 'GSE69952.Brain.RegressionInput.txt',\n",
       " 'GSE7958.Brain.RegressionInput.txt',\n",
       " 'GSE80465.Brain.RegressionInput.txt',\n",
       " 'GSE85084.Brain.RegressionInput.txt',\n",
       " 'GSE87404.Muscle.RegressionInput.txt']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "#reading the metadata column:\n",
    "metadata = pd.read_excel('/Users/andrewstapran/Desktop/Skoltech/Term_8/Computational_Biology_of_Ageing/rnaseq-project-CBA/metadata.xlsx',sheet_name='TableS1')\n",
    "metadata.columns = metadata.loc[0]\n",
    "metadata = metadata[1:]\n",
    "\n",
    "def input_list(df,Species,Technology,Tissue):\n",
    "    subset1 = df[df['Species'] == Species]\n",
    "    \n",
    "    if Technology != 'all':\n",
    "        if Technology == 'Microarray':\n",
    "            subset2 = subset1[subset1['Technology'] == 'Microarray']\n",
    "        elif Technology == 'RNAseq':\n",
    "            subset2 = subset1[subset1['Technology'] == 'RNAseq']\n",
    "    else:\n",
    "        subset2 = subset1\n",
    "    \n",
    "    if Tissue != 'all':\n",
    "        subset3 = subset2[subset2['Tissue'].isin(Tissue)]\n",
    "    else:\n",
    "        subset3 = subset2\n",
    "        \n",
    "    return list(subset3['Name_Dataset'])\n",
    "\n",
    "\n",
    "\n",
    "#mylist = input_list(metadata,'Mouse','Microarray',['Brain','Heart','Muscle'])\n",
    "mylist = input_list(metadata,'Mouse','all',['Muscle','Heart','Brain'])\n",
    "mylist"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3db35baf",
   "metadata": {},
   "source": [
    "### Controlling for samples homogeneity using PCA analysis\n",
    "This part of the code returns a PCA analysis plot based on expression of all genes present in the dataset for all samples. Initially the idea was to remove the sampples that fall outside their age cluster. However, it seemed that age groups were not as clearly separated on the PCA plot, so we just left all of them for the further analysis.<br>\n",
    "In case you will be using this code please change the paths to reporsitory with gene count values which you can find in our <a href=\"https://github.com/d-kozhevnikova/Ageing-transcriptome-meta-analysis\">GitHub</a> reporsitory. These are marked with the following comment:\n",
    "```Python\n",
    "### CHANGE THE PATH\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de7219b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#quality control:\n",
    "#function to read the input list of datasets\n",
    "def dataset_reader(input_dataset):\n",
    "    if 'AgeMap' in str(input_dataset):\n",
    "        ### CHANGE THE PATH\n",
    "        df_main = pd.read_csv(f'/Users/andrewstapran/Desktop/rnaseq-project-CBA/AgeingSignatures2020_supplementary/SupplementaryDatasets1-RegressionInputs/{input_dataset}',sep=' ',skiprows=[0,1],header=None)\n",
    "        ### CHANGE THE PATH\n",
    "        df_header = pd.read_csv(f'/Users/andrewstapran/Desktop/rnaseq-project-CBA/AgeingSignatures2020_supplementary/SupplementaryDatasets1-RegressionInputs/{input_dataset}',sep='\\t',nrows=2,header=None)\n",
    "        df = pd.concat([df_header,df_main]).reset_index(drop=True)\n",
    "    else:\n",
    "        ### CHANGE THE PATH\n",
    "        df = pd.read_csv(f'/Users/andrewstapran/Desktop/rnaseq-project-CBA/AgeingSignatures2020_supplementary/SupplementaryDatasets1-RegressionInputs/{input_dataset}',sep='\\t',header=None)\n",
    "    df.loc[0,0] = ''\n",
    "    df.loc[1,0] = 'Age'\n",
    "    df.columns = df.iloc[0]\n",
    "    df = df[1:]\n",
    "    df.set_index('',inplace=True)\n",
    "    df_transformed = df.T\n",
    "    return df\n",
    "\n",
    "#plotting function for PCA analysis results\n",
    "def plotterPCA(input_dataset):\n",
    "    my_dataset = dataset_reader(input_dataset)\n",
    "    meta = pd.DataFrame([list(my_dataset.columns),my_dataset.loc['Age']]).T\n",
    "    meta.columns = ['','Age']\n",
    "    meta.set_index('', inplace=True)\n",
    "    meta['Age']\n",
    "\n",
    "    x = np.transpose(my_dataset.values)\n",
    "    x = StandardScaler().fit_transform(x)\n",
    "    pca = PCA(n_components=2)\n",
    "    principalComponents = pca.fit_transform(x)\n",
    "    principalDf = pd.DataFrame(data = principalComponents, columns = ['PC1', 'PC2'])\n",
    "    pcdf = pd.concat([principalDf.set_index(meta.index), meta], axis = 1)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(6, 6))\n",
    "    sns.set_context(\"paper\", font_scale=1.5)\n",
    "    sns.scatterplot(data=pcdf, x=\"PC1\", y=\"PC2\", hue=\"Age\", s = 50)\n",
    "    plt.legend(loc=\"upper right\", fontsize=10)\n",
    "\n",
    "    var_pc1 = (pca.explained_variance_ratio_[0]*100).round(3)\n",
    "    var_pc2 = (pca.explained_variance_ratio_[1]*100).round(3)\n",
    "    plt.xlabel('PC1, ('+ var_pc1.astype(str) + \"%)\", fontsize = 20, labelpad = 10)\n",
    "    plt.ylabel('PC2, ('+ var_pc2.astype(str) + \"%)\", fontsize = 20, labelpad = 2)\n",
    "    plotname = input_dataset.replace('.RegressionInput.txt','').replace('.',' ')\n",
    "    plt.title(plotname, fontsize = 20)\n",
    "    ax.grid()\n",
    "\n",
    "#running the function for one of the datasets in the mylist (list of dataset names that are currently being analysed)\n",
    "plotterPCA(mylist[4])\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4aa3a35f",
   "metadata": {},
   "source": [
    "### Running linear regression for each gene for each dataset\n",
    "This cell runs the liniear regression for each gene using <a href=\"https://www.statsmodels.org/dev/generated/statsmodels.regression.linear_model.OLS.html\">statsmodels.OLS</a> function. Age values of samples in the dataset are used as predictors (X) and gene expression values for a particular gene (Y) as response variables. <i>P-values</i> and <i>standard errors</i> are obtained from the output of the same function.<br>The following function are created here:\n",
    "<ol>\n",
    "<li><b><i>return_estimators</i></b></li>\n",
    "    <ul>\n",
    "        <li><b>Takes in</b> the list of dataset names obtained in the previous section code</li>\n",
    "        <li>Reads the initial tsv file into pandas dataframe</li>\n",
    "        <li>Runs the linear regression for each gene with the dataset</li>\n",
    "        <li><b>Returns</b> the list of following parameters:</li>\n",
    "            <ol>\n",
    "                <li>Genes differentially expressed with age in this dataset</li>\n",
    "                <li>Coefficients for regression slopes for these genes</li>\n",
    "                <li>Standard errors for these coefficients</li>\n",
    "                <li>Datasets studied</li>\n",
    "                <li>Raw expression values for each gene for each dataset</li>\n",
    "                <li>Number of samples in each dataset</li>\n",
    "                <li>All genes studied in these datasets</li>\n",
    "                <li>List of % of underexpressed genes from total gene number for each dataset</li>\n",
    "                <li>List of % of overexpressed genes from total gene number for each dataset</li>\n",
    "            </ol>\n",
    "    </ul><br>\n",
    "<li><b><i>return_estimators_binomial</i></b></li>\n",
    "    <ul>\n",
    "        <li>Does similar things to what the <b><i>return_estimators</i></b> is doing, however selects only genes with statistically significant slopes for the binomial cumulative analysis</li>\n",
    "        <li><b>Returns</b> the list of following parameters:</li>\n",
    "            <ol>\n",
    "                <li>Genes differentially expressed with age in this dataset</li>\n",
    "                <li>Coefficients for regression slopes for these genes</li>\n",
    "                <li>Standard errors for these coefficients</li>\n",
    "                <li>Datasets studied</li>\n",
    "                <li>Raw expression values for each gene for each dataset</li>\n",
    "                <li>Number of samples in each dataset</li>\n",
    "            </ol>\n",
    "    </ul><br>\n",
    "<li><b><i>dataset_builder</i></b></li>\n",
    "    <ul>\n",
    "        <li><b>Takes in</b> the parameters returned by the previous two functions and constructs the datasets with columns as <i>datasets</i> and rows as <i>genes</i> - in the intersections either coefficients, standard errors or something else is given</li>\n",
    "        <li><b>Returns</b> the list of following parameters:</li>\n",
    "            <ol>\n",
    "                <li>Dataframe for regression slopes coefficients</li>\n",
    "                <li>Dataframe for standard errors for regression slopes coefficients</li>\n",
    "                <li>Dataframe for expression count values for each gene for each dataset</li>\n",
    "                <li>Sample size for each dataset</li>\n",
    "            </ol>\n",
    "    </ul><br>\n",
    "<li><b><i>binomial_estimator</i></b></li>\n",
    "    <ul>\n",
    "        <li><b>Takes in</b> the parameters returned by the <b><i>return_estimators_binomial</i></b> and <b><i>return_estimators</i></b> functions + writes union of all genes studied in selected datasets to specified file</li>\n",
    "        <li><b>Returns</b> the list of following parameters:</li>\n",
    "            <ol>\n",
    "                <li>Dictionary with genes as keys and total occurences number (in how many datasets they are studied) </li>\n",
    "                <li>Average percentage of statistically significantly underexpressed genes (from total number of genes studied)</li>\n",
    "                <li>Average percentage of statistically significantly overexpressed genes (from total number of genes studied)</li>\n",
    "            </ol>\n",
    "    </ul><br>\n",
    "</ol>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be91ffa8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#This function reads the initial tsv file into pandas dataframe\n",
    "#After this it runs the linear regression for each gene with the dataset\n",
    "def return_estimators(list_input):\n",
    "    datasets = []\n",
    "    genes = []\n",
    "    coefs = []\n",
    "    std_errs = []\n",
    "    dataset_values = []\n",
    "    sample_number_list = []\n",
    "    all_genes_studied = []\n",
    "    probabilities_under = []\n",
    "    probabilities_over = []\n",
    "    \n",
    "    for each in list_input:\n",
    "        print(each)\n",
    "        datasets.append(each)\n",
    "        if 'AgeMap' in str(each):\n",
    "            ### CHANGE THE PATH\n",
    "            df_main = pd.read_csv(f'/Users/andrewstapran/Desktop/rnaseq-project-CBA/AgeingSignatures2020_supplementary/SupplementaryDatasets1-RegressionInputs/{each}',sep=' ',skiprows=[0,1],header=None)\n",
    "            ### CHANGE THE PATH\n",
    "            df_header = pd.read_csv(f'/Users/andrewstapran/Desktop/rnaseq-project-CBA/AgeingSignatures2020_supplementary/SupplementaryDatasets1-RegressionInputs/{each}',sep='\\t',nrows=2,header=None)\n",
    "            df = pd.concat([df_header,df_main]).reset_index(drop=True)\n",
    "        else:\n",
    "            ### CHANGE THE PATH\n",
    "            df = pd.read_csv(f'/Users/andrewstapran/Desktop/rnaseq-project-CBA/AgeingSignatures2020_supplementary/SupplementaryDatasets1-RegressionInputs/{each}',sep='\\t',header=None)\n",
    "        df.loc[0,0] = ''\n",
    "        df.loc[1,0] = 'Age'\n",
    "        df.columns = df.iloc[0]\n",
    "        df = df[1:]\n",
    "        df.set_index('',inplace=True)\n",
    "        sample_number = len(list(df.columns))\n",
    "        sample_number_list.append(sample_number)\n",
    "        raw_values = ['age'] + list(df.values.tolist()[1:])\n",
    "        df_transformed = df.T\n",
    "        x = np.array(list(df_transformed['Age'].astype(float)))#.reshape((-1, 1))\n",
    "        list_slopes = ['age']\n",
    "        p_values = ['age']\n",
    "        std_errs_list = ['age']\n",
    "        \n",
    "        for each in list(df_transformed.columns)[1::]:\n",
    "            y = np.array(list(df_transformed[each].astype(float)))\n",
    "            X = sm.add_constant(x)\n",
    "            model = sm.OLS(y,X).fit()\n",
    "            p_value = model.pvalues[-1]\n",
    "            std_err = model.bse[-1]\n",
    "            coef = model.params[-1]\n",
    "            \n",
    "            list_slopes.append(coef)\n",
    "            p_values.append(p_value)\n",
    "            std_errs_list.append(std_err)\n",
    "            \n",
    "        df_transformed.loc['slopes'] = list_slopes\n",
    "        df_transformed.loc['pvalues'] = p_values\n",
    "        df_transformed.loc['std_errs'] = std_errs_list\n",
    "        df_transformed.loc['values'] = raw_values\n",
    "        \n",
    "        df_final = df_transformed.T\n",
    "        all_genes_studied.append(list(df_final.index)[1:])\n",
    "        df_final2 = df_final[df_final['pvalues'] != 'age']\n",
    "        df_for_binomial = df_final2[df_final2['pvalues'] < 0.05]\n",
    "        df_final_stat_sign = df_final2\n",
    "        \n",
    "        genes.append(list(df_final_stat_sign.index))\n",
    "        coefs.append(list(df_final_stat_sign['slopes']))\n",
    "        std_errs.append(list(df_final_stat_sign['std_errs']))\n",
    "        dataset_values.append(list(df_final_stat_sign['values']))\n",
    "        \n",
    "        total_genes_number = len(list(df_final.index)[1:])\n",
    "        underexp_number = len(df_for_binomial[df_for_binomial['slopes'] < 0])\n",
    "        print(f'Total number of genes underexpressed with age: {underexp_number}')\n",
    "        overexp_number = len(df_for_binomial[df_for_binomial['slopes'] > 0])\n",
    "        print(f'Total number of genes overexpressed with age: {overexp_number}')\n",
    "        probabilities_under.append(float(underexp_number / total_genes_number))\n",
    "        probabilities_over.append(float(overexp_number / total_genes_number))\n",
    "        print('---------------')\n",
    "        \n",
    "    \n",
    "    #dataset = dataset_builder([genes,coefs,datasets])\n",
    "    return [genes,coefs,std_errs,datasets,dataset_values,\n",
    "            sample_number_list,all_genes_studied,\n",
    "            probabilities_under,probabilities_over,]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def return_estimators_binomial(list_input):\n",
    "    datasets = []\n",
    "    genes = []\n",
    "    coefs = []\n",
    "    std_errs = []\n",
    "    dataset_values = []\n",
    "    sample_number_list = []\n",
    "    all_genes_studied = []\n",
    "    probabilities_under = []\n",
    "    probabilities_over = []\n",
    "    \n",
    "    for each in list_input:\n",
    "        print(each)\n",
    "        datasets.append(each)\n",
    "        if 'AgeMap' in str(each):\n",
    "            ### CHANGE THE PATH\n",
    "            df_main = pd.read_csv(f'/Users/andrewstapran/Desktop/rnaseq-project-CBA/AgeingSignatures2020_supplementary/SupplementaryDatasets1-RegressionInputs/{each}',sep=' ',skiprows=[0,1],header=None)\n",
    "            ### CHANGE THE PATH\n",
    "            df_header = pd.read_csv(f'/Users/andrewstapran/Desktop/rnaseq-project-CBA/AgeingSignatures2020_supplementary/SupplementaryDatasets1-RegressionInputs/{each}',sep='\\t',nrows=2,header=None)\n",
    "            df = pd.concat([df_header,df_main]).reset_index(drop=True)\n",
    "        else:\n",
    "            ### CHANGE THE PATH\n",
    "            df = pd.read_csv(f'/Users/andrewstapran/Desktop/rnaseq-project-CBA/AgeingSignatures2020_supplementary/SupplementaryDatasets1-RegressionInputs/{each}',sep='\\t',header=None)\n",
    "        df.loc[0,0] = ''\n",
    "        df.loc[1,0] = 'Age'\n",
    "        df.columns = df.iloc[0]\n",
    "        df = df[1:]\n",
    "        df.set_index('',inplace=True)\n",
    "        sample_number = len(list(df.columns))\n",
    "        sample_number_list.append(sample_number)\n",
    "        raw_values = ['age'] + list(df.values.tolist()[1:])\n",
    "        df_transformed = df.T\n",
    "        x = np.array(list(df_transformed['Age'].astype(float)))#.reshape((-1, 1))\n",
    "        list_slopes = ['age']\n",
    "        p_values = ['age']\n",
    "        std_errs_list = ['age']\n",
    "        \n",
    "        for each in list(df_transformed.columns)[1::]:\n",
    "            y = np.array(list(df_transformed[each].astype(float)))\n",
    "            X = sm.add_constant(x)\n",
    "            model = sm.OLS(y,X).fit()\n",
    "            p_value = model.pvalues[-1]\n",
    "            std_err = model.bse[-1]\n",
    "            coef = model.params[-1]\n",
    "            \n",
    "            list_slopes.append(coef)\n",
    "            p_values.append(p_value)\n",
    "            std_errs_list.append(std_err)\n",
    "            \n",
    "        df_transformed.loc['slopes'] = list_slopes\n",
    "        df_transformed.loc['pvalues'] = p_values\n",
    "        df_transformed.loc['std_errs'] = std_errs_list\n",
    "        df_transformed.loc['values'] = raw_values\n",
    "        \n",
    "        df_final = df_transformed.T\n",
    "        all_genes_studied.append(list(df_final.index)[1:])\n",
    "        df_final2 = df_final[df_final['pvalues'] != 'age']\n",
    "        df_final_stat_sign = df_final2[df_final2['pvalues'] < 0.05]\n",
    "        \n",
    "        genes.append(list(df_final_stat_sign.index))\n",
    "        coefs.append(list(df_final_stat_sign['slopes']))\n",
    "        std_errs.append(list(df_final_stat_sign['std_errs']))\n",
    "        dataset_values.append(list(df_final_stat_sign['values']))\n",
    "        \n",
    "        total_genes_number = len(list(df_final.index)[1:])\n",
    "        underexp_number = len(df_final_stat_sign[df_final_stat_sign['slopes'] < 0])\n",
    "        print(f'Total number of genes underexpressed with age: {underexp_number}')\n",
    "        overexp_number = len(df_final_stat_sign[df_final_stat_sign['slopes'] > 0])\n",
    "        print(f'Total number of genes overexpressed with age: {overexp_number}')\n",
    "        probabilities_under.append(float(underexp_number / total_genes_number))\n",
    "        probabilities_over.append(float(overexp_number / total_genes_number))\n",
    "        print('---------------')\n",
    "        \n",
    "    \n",
    "    #dataset = dataset_builder([genes,coefs,datasets])\n",
    "    return [genes,coefs,std_errs,datasets,dataset_values,\n",
    "            sample_number_list]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#combining linear regression coefficients, standard errors from different datasets into one dataframe\n",
    "def dataset_builder(input_coefs_genes):\n",
    "    input_genes = input_coefs_genes[0]\n",
    "    input_coefs = input_coefs_genes[1]\n",
    "    input_stderrs = input_coefs_genes[2]\n",
    "    input_datasets = input_coefs_genes[3]\n",
    "    input_values = input_coefs_genes[4]\n",
    "    input_sample_size = input_coefs_genes[5]\n",
    "    \n",
    "    dataset_name = input_datasets[0].replace('.RegressionInput.txt','')\n",
    "    initial_coefs = pd.DataFrame([input_genes[0],input_coefs[0]]).T\n",
    "    initial_stderrs = pd.DataFrame([input_genes[0],input_stderrs[0]]).T\n",
    "    initial_values = pd.DataFrame([input_genes[0],input_values[0]]).T\n",
    "    \n",
    "    initial_coefs.columns = ['Gene',dataset_name]\n",
    "    initial_stderrs.columns = ['Gene',dataset_name]\n",
    "    initial_values.columns = ['Gene',dataset_name]\n",
    "    \n",
    "    initial_coefs['Gene'] = initial_coefs['Gene'].astype(int)\n",
    "    initial_stderrs['Gene'] = initial_stderrs['Gene'].astype(int)\n",
    "    initial_values['Gene'] = initial_values['Gene'].astype(int)\n",
    "    \n",
    "    for index, element in enumerate(input_genes[1:]):\n",
    "        dataset_name = input_datasets[index+1].replace('.RegressionInput.txt','')\n",
    "        to_add_coefs = pd.DataFrame([element,input_coefs[index+1]]).T\n",
    "        to_add_stderrs = pd.DataFrame([element,input_stderrs[index+1]]).T\n",
    "        to_add_values = pd.DataFrame([element,input_values[index+1]]).T\n",
    "        \n",
    "        to_add_coefs.columns = ['Gene',dataset_name]\n",
    "        to_add_stderrs.columns = ['Gene',dataset_name]\n",
    "        to_add_values.columns = ['Gene',dataset_name]\n",
    "        \n",
    "        to_add_coefs['Gene'] = to_add_coefs['Gene'].astype(int)\n",
    "        to_add_stderrs['Gene'] = to_add_stderrs['Gene'].astype(int)\n",
    "        to_add_values['Gene'] = to_add_values['Gene'].astype(int)\n",
    "        \n",
    "        initial_coefs = pd.merge(initial_coefs,to_add_coefs,on='Gene',how='outer').fillna('not_pres')\n",
    "        initial_stderrs = pd.merge(initial_stderrs,to_add_stderrs,on='Gene',how='outer').fillna('not_pres')\n",
    "        initial_values = pd.merge(initial_values,to_add_values,on='Gene',how='outer').fillna('not_pres')\n",
    "        \n",
    "    initial_coefs.set_index('Gene',inplace=True)\n",
    "    initial_stderrs.set_index('Gene',inplace=True)\n",
    "    initial_values.set_index('Gene',inplace=True)\n",
    "    return [initial_coefs,initial_stderrs,initial_values,input_sample_size]\n",
    "        \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def binomial_estimator(input_result,total_studied_path):\n",
    "    list_all_genes = input_result[6]\n",
    "    final_list = []\n",
    "    for element in list_all_genes:\n",
    "        final_list = final_list + [str(i) for i in element]\n",
    "    print(f'Total number of genes: {len(final_list)}')\n",
    "    final_list_unique = list(set(final_list))\n",
    "    total_genes = open(total_studied_path,'w')\n",
    "    for gene_element in final_list_unique:\n",
    "        total_genes.write(f'{gene_element}\\n')\n",
    "    total_genes.close()\n",
    "    print(f'Total number of unique genes: {len(final_list_unique)}')\n",
    "    gene_occurences = {}\n",
    "    for element in final_list_unique:\n",
    "        print(f'{final_list_unique.index(element)} out of {len(final_list_unique)}')\n",
    "        counter = 0\n",
    "        for each in list_all_genes:\n",
    "            new_list = [str(bublik) for bublik in each]\n",
    "            if element in new_list:\n",
    "                counter += 1\n",
    "        gene_occurences[element] = counter\n",
    "    \n",
    "    average_under_P = sum(input_result[7]) / len(input_result[7])\n",
    "    average_over_P = sum(input_result[8]) / len(input_result[8])\n",
    "    return [gene_occurences, average_under_P, average_over_P]\n",
    "\n",
    "\n",
    "\n",
    "#running the functions\n",
    "a = return_estimators(mylist)        \n",
    "mydata = dataset_builder(a)\n",
    "### CHANGE THE PATH\n",
    "general_stats_on_selected_datasets = binomial_estimator(a,'/Users/andrewstapran/Desktop/mygenesall.txt')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "748ff632",
   "metadata": {},
   "source": [
    "### Subsetting the dataframe for datasets\n",
    "In the previous step we have obtained the dataset with expression-age regression slope coefficients for each gene for each dataset. Now we need to select genes that are associated with age in <b><i>at least 2 datasets</i></b>.<br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d9a66bd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def subsetting_stat_genes(df):\n",
    "    #finding genes that were studied in at least 2 datasets\n",
    "    dataset_number = len(df[0].columns)\n",
    "    # if dataset_number % 2 == 0:\n",
    "    #     threshold = dataset_number / 2\n",
    "    # else:\n",
    "    #     threshold = int(dataset_number / 2) + 1\n",
    "\n",
    "    coefs = df[0]\n",
    "    stderrs = df[1]\n",
    "    values = df[2]\n",
    "    sample_size = df[3]\n",
    "\n",
    "    genes_studied = list(coefs.index)\n",
    "    genes_to_study = []\n",
    "    for element in genes_studied:\n",
    "        number_abs = list(coefs.loc[element]).count('not_pres')\n",
    "        if int(number_abs) <= dataset_number - 2:\n",
    "            genes_to_study.append(element)\n",
    "        else:\n",
    "            pass\n",
    "\n",
    "    coefs_atleast = coefs.loc[genes_to_study]    \n",
    "    stderrs_atleast = stderrs.loc[genes_to_study]\n",
    "    values_atleast = values.loc[genes_to_study]\n",
    "    return [coefs_atleast,stderrs_atleast,values_atleast,sample_size]\n",
    "\n",
    "myraw = subsetting_stat_genes(mydata)\n",
    "myraw_binomial = subsetting_stat_genes(mydata_binomial)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "040fa3ec",
   "metadata": {},
   "source": [
    "<hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "728552d8",
   "metadata": {},
   "source": [
    "### Main code:<br>Running <b><i>metaregression</i></b> analysis\n",
    "There are two function below for running the meta-regression analysis:\n",
    "<ol>\n",
    "    <li><i><b>gene_list_return</b></i> using PyMare meta-regression</li>\n",
    "    <li><i><b>binomial_check</b></i> using Binomial Cumulative test meta-regression</li>\n",
    "</ol>\n",
    "Both tests use multiple test correction for p-value adjustment using the <a href=\"https://www.statsmodels.org/dev/generated/statsmodels.stats.multitest.fdrcorrection.html\">statsmodels.stats.multitest.fdrcorrection</a> function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47e16b3d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def not_pres_remove(input_list):\n",
    "    final_list = [element for element in input_list if element != 'not_pres']\n",
    "    return final_list\n",
    "\n",
    "enc = preprocessing.OneHotEncoder()\n",
    "a = np.array(['Heart','Muscle','Heart','Brain','Heart']).reshape(-1, 1)\n",
    "b = enc.fit(a)\n",
    "\n",
    "def gene_list_return(input_data):\n",
    "    coefs_pymare = input_data[0].copy()\n",
    "    column_names = [each.split('.')[-1] for each in list(coefs_pymare.columns)]\n",
    "    column_names_corr = enc.transform(np.array(column_names).reshape(-1, 1)).toarray()\n",
    "    \n",
    "    variances = input_data[1].copy()\n",
    "    dataset_size = input_data[3].copy()\n",
    "    p_values = []\n",
    "    coefs_final = []\n",
    "    for element in list(coefs_pymare.index):\n",
    "        print(f'{element} --> {list(coefs_pymare.index).index(element)} out of {len(list(coefs_pymare.index))}')\n",
    "        initial_elements = list(coefs_pymare.loc[element])\n",
    "        y = not_pres_remove(list(coefs_pymare.loc[element]))\n",
    "        y_num = [float(each) for each in y]\n",
    "        v = not_pres_remove(list(variances.loc[element]))\n",
    "        v_num = [float(each) for each in v]\n",
    "        v_squared = [i ** 2 for i in v_num]\n",
    "        omit_list = [index for index, element in enumerate(initial_elements) if element == 'not_pres']\n",
    "        study_list_size = [element for index, element in enumerate(dataset_size) if index not in omit_list]\n",
    "        column_names_list = [each for index1, each in enumerate(column_names_corr) if index1 not in omit_list]\n",
    "        model = core.meta_regression(y = y_num, v = v_squared, X = column_names_list)\n",
    "        #model = core.meta_regression(y = y_num, v = v_squared, n = study_list_size)\n",
    "        final_df = model.to_df(alpha=0.05)\n",
    "        p_value = final_df.loc[0,'p-value']\n",
    "        estimate = final_df.loc[0,'estimate']\n",
    "        p_values.append(p_value)\n",
    "        coefs_final.append(estimate)\n",
    "    coefs_pymare['total_effect'] = coefs_final\n",
    "    coefs_pymare['total_p_values'] = p_values\n",
    "    coefs_pymare_subset = coefs_pymare[coefs_pymare['total_p_values'].notna()]\n",
    "    coefs_pymare_subset['adjusted_p_values'] = list(statsmodels.stats.multitest.fdrcorrection(list(coefs_pymare_subset['total_p_values']),alpha=0.05,is_sorted=False)[1])\n",
    "    #coefs_pymare['adjusted_p_values'] = list(statsmodels.stats.multitest.fdrcorrection(p_values,alpha=0.05,is_sorted=False)[1])\n",
    "    return coefs_pymare_subset\n",
    "\n",
    "\n",
    "\n",
    "#binomial test metaregression:\n",
    "def binomial_check(input_data, general_stats_input):\n",
    "    coefs_binom2 = input_data[0].copy()\n",
    "    genes_left = [int(gene) for gene in list(final_meta_df.index)]\n",
    "    intersection = [int(gene_ID) for gene_ID in genes_left if int(gene_ID) in list(coefs_binom2.index)]\n",
    "    coefs_binom = coefs_binom2.loc[intersection]\n",
    "    print(len(coefs_binom))\n",
    "    datasets = list(coefs_binom.columns)\n",
    "    print(datasets)\n",
    "    total_occurences = general_stats_input[0]\n",
    "    under_P_average = general_stats_input[1]\n",
    "    over_P_average = general_stats_input[2]\n",
    "    binomial_Ps = []\n",
    "    for each in list(coefs_binom.index):\n",
    "        list_row = [float(element) for element in list(coefs_binom.loc[each]) if element != 'not_pres']\n",
    "        k_pos = len([item for item in list_row if float(item) > 0])\n",
    "        k_neg = len([item for item in list_row if float(item) < 0])\n",
    "        if k_pos > k_neg:\n",
    "            probability = over_P_average\n",
    "            k = k_pos\n",
    "        elif k_pos < k_neg:\n",
    "            probability = under_P_average\n",
    "            k = k_neg\n",
    "        else:\n",
    "            k = k_pos\n",
    "            effect = final_meta_df.loc[each,'total_effect']\n",
    "            if effect > 0:\n",
    "                probability = over_P_average\n",
    "            elif effect < 0:\n",
    "                probability = under_P_average\n",
    "        n = total_occurences[str(each)]\n",
    "        binomial_Ps.append(1 - scipy.stats.binom.cdf(k, n, probability, loc=0))\n",
    "    final_df = pd.DataFrame([[str(each) for each in list(coefs_binom.index)],binomial_Ps]).T\n",
    "    final_df.columns = ['Genes','p_value']\n",
    "    final_df.set_index('Genes', inplace=True)\n",
    "    p_values = [float(element) for element in list(final_df['p_value'])]\n",
    "    final_df['adjusted_p_value'] = list(statsmodels.stats.multitest.fdrcorrection(p_values,alpha=0.05,is_sorted=False)[1])\n",
    "    return final_df\n",
    "\n",
    "#running the meta-regression analysis\n",
    "final_meta_df = gene_list_return(myraw)\n",
    "binomial_results = binomial_check(myraw_binomial, general_stats_on_selected_datasets)\n",
    "subset_final_pymare = final_meta_df.loc[[int(each) for each in list(binomial_results.index)]]\n",
    "binomial_results['effect_meta'] = list(subset_final_pymare['total_effect'])\n",
    "\n",
    "\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "655997b3",
   "metadata": {},
   "source": [
    "### Comparing the results of PyMare and Binomial meta-regression:\n",
    "This part of the code plots the results of PyMare meta-regression against Binomial meta-regression. Each approach gives a specific p-value to each gene based on how frequently stastically signifincant association of this gene is observed in the studied datasets. Therefore we can rank these genes from based on their p-values (starting with best up-regulated genes and ending with best down-regulated genes). And then plot the rank positions of the genes commonly identified by both analysis types, calculating the correlaiton coefficient. This is returned as output of the function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b80bf463",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def group_pos(input_df,typeofdata):\n",
    "    positions = list(range(0,len(input_df),1))\n",
    "    input_df['index_pos'] = positions\n",
    "    if typeofdata == 'binomial':\n",
    "        input_df_grouped = input_df.groupby('adjusted_p_value')['index_pos'].mean().reset_index()\n",
    "        input_df_dict = dict(zip(list(input_df_grouped.adjusted_p_value),list(input_df_grouped.index_pos)))\n",
    "        final_pos = dict(zip([int(each) for each in list(input_df.index)],[input_df_dict[each] for each in list(input_df.adjusted_p_value)]))\n",
    "    elif typeofdata == 'pymare':\n",
    "        input_df_grouped = input_df.groupby('adjusted_p_values')['index_pos'].mean().reset_index()\n",
    "        input_df_dict = dict(zip(list(input_df_grouped.adjusted_p_values),list(input_df_grouped.index_pos)))\n",
    "        final_pos = dict(zip([int(each) for each in list(input_df.index)],[int(input_df_dict[each]) for each in list(input_df.adjusted_p_values)]))\n",
    "    return final_pos\n",
    "\n",
    "def plotter(binomial_input,pymare_input):\n",
    "    binomial_input_nonzero = binomial_input[binomial_input['adjusted_p_value'] != 0]\n",
    "    list_genes_binomial = [int(each) for each in list(binomial_input_nonzero.index)]\n",
    "    pymare_input_nonzero = pymare_input[pymare_input['adjusted_p_values'] != 0]\n",
    "    list_genes_pymare = [int(each) for each in list(pymare_input_nonzero.index)]\n",
    "    \n",
    "    common_list_binomial = [str(each) for each in list_genes_binomial if each in list_genes_pymare]\n",
    "    common_list_pymare = [int(each) for each in list_genes_binomial if each in list_genes_pymare]\n",
    "    binomial_subset = binomial_input_nonzero.loc[common_list_binomial]\n",
    "    pymare_subset = pymare_input_nonzero.loc[common_list_pymare]\n",
    "    \n",
    "    pymare_over_subset = pymare_subset[pymare_subset['total_effect'] > 0]\n",
    "    pymare_over_subset.sort_values(by=['adjusted_p_values'],ascending=True,inplace=True)\n",
    "    pymare_under_subset = pymare_subset[pymare_subset['total_effect'] < 0]\n",
    "    pymare_under_subset.sort_values(by=['adjusted_p_values'],ascending=False,inplace=True)\n",
    "    binomial_over_subset = binomial_subset[binomial_subset['effect_meta'] > 0]\n",
    "    binomial_over_subset.sort_values(by=['adjusted_p_value'],ascending=True,inplace=True)\n",
    "    binomial_under_subset = binomial_subset[binomial_subset['effect_meta'] < 0]\n",
    "    binomial_under_subset.sort_values(by=['adjusted_p_value'],ascending=False,inplace=True)\n",
    "    pymare_final = pd.concat([pymare_over_subset,pymare_under_subset])\n",
    "    binomial_final = pd.concat([binomial_over_subset,binomial_under_subset])\n",
    "    #pymare_final = pymare_under_subset\n",
    "    #binomial_final = binomial_under_subset\n",
    "    \n",
    "    list_pymare = group_pos(pymare_final,'pymare')\n",
    "    list_binomial = group_pos(binomial_final,'binomial')\n",
    "#     list_pymare = [str(each) for each in list(pymare_final.index)]\n",
    "#     list_binomial = [str(each) for each in list(binomial_final.index)]\n",
    "    resulting_dict = {}\n",
    "    for element in common_list_pymare:\n",
    "        #element = str(element2)\n",
    "        index_pymare = list_pymare[element]\n",
    "        #index_pymare = list_pymare.index(element)\n",
    "        index_binomial = list_binomial[element]\n",
    "        #index_binomial = list_binomial.index(element)\n",
    "        resulting_dict[index_pymare] = index_binomial\n",
    "    plt.scatter(list(resulting_dict.keys()),list(resulting_dict.values()))\n",
    "    plt.xlabel('PyMare analysis based rank', fontsize=15)\n",
    "    plt.ylabel('Cumulative binomial test\\nbased rank', fontsize=15)\n",
    "    res = scipy.stats.spearmanr(list(resulting_dict.keys()), list(resulting_dict.values()), axis=0,alternative='two-sided')\n",
    "    print(res)\n",
    "    plt.text(max(list(resulting_dict.keys()))*0.5,max(list(resulting_dict.values()))*0.1,f'Spearman:\\nCorrelation: {res.correlation:.4f}\\nP-value: {res.pvalue:.4f}',backgroundcolor = 'lightgray')    \n",
    "    #plt.text(5000,5000,res.correlation,backgroundcolor = 'gray')    \n",
    "    plt.show()\n",
    "    return [resulting_dict,pymare_final,binomial_final]\n",
    "\n",
    "a = plotter(binomial_results,final_meta_df)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "619ff9e0",
   "metadata": {},
   "source": [
    "### Converting Entrez_ID to Gene_IDs\n",
    "Original datasets have been using the Entrez_ID identifiers for the genes which look something like this: <b><i>103018</i></b>. However using <a href=\"https://pypi.org/project/mygene/\">mygene</a> package we converted them to conventional gene names, like <b><i>APOE</i></b>. Afterwards, the resulting dataframe containing:\n",
    "<ul>\n",
    "    <li>Coefficients for each gene for each dataset</li>\n",
    "    <li>EntrezID</li>\n",
    "    <li>Conventional gene names</li>\n",
    "    <li>Total effect column - output of the PyMare meta-regression function (somthing like an \"average\" regression coefficient for all datasets for this gene). If positive - upregulated with age, if negative - vice versa</li>\n",
    "    <li>Raw p-values for total effect estimates</li>\n",
    "    <li>Benjamini-Hochberg FDR adjusted p-values for total effect estimates</li>\n",
    "</ul>\n",
    "Importantly, some Entrez_ID correspond to genes that have either been discontinued from using in annotation or to some putative protein-coding regions. For such Entrez_ID mygene package is unable to find the conventional gene name - these are replaced with the <b><i>geneID_not_found</i></b> label. It is important to weed them out from gene lists before running the further analysis steps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "146b8b3b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "mg = mygene.MyGeneInfo()\n",
    "identificator = 'threetissues_Xconsidering'\n",
    "def list_gene_name_converter(input_gene_list):\n",
    "    list_entrez_gene_IDs = [int(each) for each in input_gene_list]\n",
    "    output = mg.getgenes(list_entrez_gene_IDs,fields='symbol')\n",
    "    list_geneIDs = []\n",
    "    for element in output:\n",
    "        if 'symbol' in list(element.keys()):\n",
    "            list_geneIDs.append(element['symbol'].upper())\n",
    "        else:\n",
    "            list_geneIDs.append('geneID_not_found')\n",
    "    return list_geneIDs\n",
    "\n",
    "def convert_gene_names(input_df):\n",
    "    list_gene_IDs = list_gene_name_converter(list(input_df.index))\n",
    "    return_df = input_df.copy()\n",
    "    return_df['gene_name'] = list_gene_IDs\n",
    "    return_df['mouse_entrez_id'] = list(return_df.index)\n",
    "    cols = return_df.columns.tolist()\n",
    "    cols = cols[-2:] + cols[:-2]\n",
    "    return_df = return_df[cols] \n",
    "    return return_df\n",
    "\n",
    "#running the function for PyMare results\n",
    "meta_pymare = convert_gene_names(final_meta_df)\n",
    "### CHANGE THE PATH\n",
    "meta_pymare.to_csv(f'/Users/andrewstapran/Desktop/{identificator}_pymare.csv',sep='\\t',header=True,index=False)\n",
    "\n",
    "#similarly we can obtain similar table with binomial cumulative testing and run table creation for these\n",
    "meta_binomial = convert_gene_names(final_df)\n",
    "### CHANGE THE PATH\n",
    "meta_binomial.to_csv(f'/Users/andrewstapran/Desktop/{identificator}_binomial.csv',sep='\\t',header=True,index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f006f4ba",
   "metadata": {},
   "source": [
    "### Converting gene list to names of genes\n",
    "[Section1](#myID1)\n",
    "As you might remember we also saved list of all genes studied across selected datasets - however they are given as Entrez_ID numbers as well. This function converts gene numbers to conventional gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "355d4a93",
   "metadata": {},
   "outputs": [],
   "source": [
    "mg = mygene.MyGeneInfo()\n",
    "def list_gene_name_converter(input_gene_list):\n",
    "    list_entrez_gene_IDs = [int(each) for each in input_gene_list]\n",
    "    output = mg.getgenes(list_entrez_gene_IDs,fields='symbol')\n",
    "    list_geneIDs = []\n",
    "    for element in output:\n",
    "        if 'symbol' in list(element.keys()):\n",
    "            list_geneIDs.append(element['symbol'].upper())\n",
    "        else:\n",
    "            gene_ID_number = element['query']\n",
    "            list_geneIDs.append(f'{gene_ID_number} - geneID_not_found')\n",
    "    return list_geneIDs\n",
    "\n",
    "list1 = []\n",
    "### CHANGE THE PATH\n",
    "with open('/Users/andrewstapran/Desktop/mygenesall.txt','r') as file1:\n",
    "    lines = file1.readlines()\n",
    "for element in lines:\n",
    "    list1.append(element.replace('\\n',''))\n",
    "file1.close()\n",
    "list2 = list_gene_name_converter(list1)\n",
    "\n",
    "### CHANGE THE PATH\n",
    "file2 = open('/Users/andrewstapran/Desktop/all_three_tissues_names.txt','w')\n",
    "for element in list2:\n",
    "    file2.write(f'{element}\\n')\n",
    "file2.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f45effc8",
   "metadata": {},
   "source": [
    "<hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f8b0914",
   "metadata": {},
   "source": [
    "### Main code:<br>Running the enrichGO analysis\n",
    "Using this bit of code we ran the enrichment analysis for <a href=\"http://geneontology.org/\">GO (Gene Ontology)</a> terms for up- and downregulated genes for each tissue as well as for meta-analysis of three tissues together. As input for the <b><i>gseapy.enrichr</i></b> function following lists are used:\n",
    "<ol>\n",
    "    <li>Gene list of statistically significantly up- or downregulated with age genes (we used <i>Gene</i> column from <i>csv</i> tables we have obtained in the previous step)</li>\n",
    "    <li>As background genes: list of genes studied across all studied datasets in this analysis which can be obtained from the previous step <i>txt</i> tables we have obtained in the previous step</li>\n",
    "</ol>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e124075d",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "### CHANGE THE PATH\n",
    "brain_table = '/Users/andrewstapran/Desktop/rnaseq-project-CBA/results/brain/brain_pymare.csv'\n",
    "brain_file = '/Users/andrewstapran/Desktop/rnaseq-project-CBA/results/brain/brain_all_genes_names.txt'\n",
    "heart_table = '/Users/andrewstapran/Desktop/rnaseq-project-CBA/results/heart/heart_pymare.csv'\n",
    "heart_file = '/Users/andrewstapran/Desktop/rnaseq-project-CBA/results/heart/heart_all_genes_names.txt'\n",
    "muscle_table = '/Users/andrewstapran/Desktop/rnaseq-project-CBA/results/muscle/muscle_pymare.csv'\n",
    "muscle_file = '/Users/andrewstapran/Desktop/rnaseq-project-CBA/results/muscle/muscle_all_genes_names.txt'\n",
    "total_table = '/Users/andrewstapran/Desktop/rnaseq-project-CBA/results/allthree/threetissues_Xconsidering_pymare.csv'\n",
    "total_file = '/Users/andrewstapran/Desktop/rnaseq-project-CBA/results/allthree/all_three_tissues_names.txt'\n",
    "\n",
    "#reading the dataframes\n",
    "def reader(input_df_path,input_file_path):\n",
    "    input_df = pd.read_csv(input_df_path,sep='\\t')\n",
    "    with open(input_file_path,'r') as file1:\n",
    "        lines = file1.readlines()\n",
    "    list_all_genes = [each.replace('\\n','') for each in lines if 'not_found' not in each]\n",
    "    stat_sign = input_df[input_df.adjusted_p_values < 0.05]\n",
    "    stat_sign_over = stat_sign[stat_sign['total_effect'] > 0]\n",
    "    stat_sign_under = stat_sign[stat_sign['total_effect'] < 0]\n",
    "    stat_sign_over_final = [each for each in list(stat_sign_over['gene_name']) if 'not_found' not in each]\n",
    "    stat_sign_under_final = [each for each in list(stat_sign_under['gene_name']) if 'not_found' not in each]\n",
    "    return [stat_sign_over_final,stat_sign_under_final,list_all_genes]\n",
    "\n",
    "\n",
    "#creating the lists of genes for each tissue\n",
    "brain_over = reader(brain_table,brain_file)[0]\n",
    "print(len(brain_over))\n",
    "brain_under = reader(brain_table,brain_file)[1]\n",
    "print(len(brain_under))\n",
    "brain_total = reader(brain_table,brain_file)[2]\n",
    "print(len(brain_total))\n",
    "heart_over = reader(heart_table,heart_file)[0]\n",
    "print(len(heart_over))\n",
    "heart_under = reader(heart_table,heart_file)[1]\n",
    "print(len(heart_under))\n",
    "heart_total = reader(heart_table,heart_file)[2]\n",
    "muscle_over = reader(muscle_table,muscle_file)[0]\n",
    "muscle_under = reader(muscle_table,muscle_file)[1]\n",
    "muscle_total = reader(muscle_table,muscle_file)[2]\n",
    "three_over = reader(total_table,total_file)[0]\n",
    "three_under = reader(total_table,total_file)[1]\n",
    "three_total = reader(total_table,total_file)[2]\n",
    "\n",
    "#part for the GO\n",
    "def plot_GO(inputlist_diff,inputlist_total,title):\n",
    "    enr = gseapy.enrichr(gene_list = inputlist_diff,\n",
    "                         background = inputlist_total,\n",
    "                         gene_sets=['GO_Biological_Process_2021', \n",
    "                                    'GO_Cellular_Component_2021', \n",
    "                                    'GO_Molecular_Function_2021'],\n",
    "                         organism='mouse',\n",
    "                         outdir=None)\n",
    "    enr_res = enr.results.iloc[:,[0,1,2,4,9]]\n",
    "    enr_res = enr_res[enr_res['Adjusted P-value'] < 0.05]\n",
    "    enr_res\n",
    "    dotplot(enr_res,\n",
    "            column=\"Adjusted P-value\",\n",
    "            x='Gene_set',\n",
    "            size=10,\n",
    "            top_term=5,\n",
    "            figsize=(3,5),\n",
    "            title = title,\n",
    "            thresh=0.5,\n",
    "            xticklabels_rot=45, \n",
    "            show_ring=True, \n",
    "            marker='o')\n",
    "\n",
    "#running the plotting functions    \n",
    "plot_GO(brain_over,brain_total,'OverExpressed in brain')\n",
    "plot_GO(brain_under,brain_total,'UnderExpressed in brain') \n",
    "plot_GO(muscle_over,muscle_total,'OverExpressed in muscle')\n",
    "plot_GO(muscle_under,muscle_total,'UnderExpressed in muscle')     \n",
    "plot_GO(heart_over,heart_total,'OverExpressed in heart')\n",
    "plot_GO(heart_under,heart_total,'UnderExpressed in heart')    \n",
    "plot_GO(three_over,three_total,'OverExpressed in All three tissues')\n",
    "plot_GO(three_under,three_total,'UnderExpressed in All three tissues')   \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4242b5ea",
   "metadata": {},
   "source": [
    "### Building the venn diagrams for gene list intersections\n",
    "This final section of the code intersects gene lists from different tissues with each other and builds venn diagrams for these intersections.<br>It also builds an intersection for genes up- and downregulated with age identified in meta-analysis of three tissues with GenAge gene list for mice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d58a2877",
   "metadata": {},
   "outputs": [],
   "source": [
    "def intersector(list1,list2):\n",
    "    list3 = [each for each in list1 if each in list2]\n",
    "    return list3\n",
    "\n",
    "#OVER:\n",
    "brain_heart_over = intersector(brain_over,heart_over)\n",
    "brain_muscle_over = intersector(brain_over,muscle_over)\n",
    "heart_muscle_over = intersector(muscle_over,heart_over)\n",
    "all_three_over = intersector(brain_heart_over,muscle_over)\n",
    "final_intersect = intersector(all_three_over, three_over)\n",
    "\n",
    "#UNDER:\n",
    "brain_heart_under = intersector(brain_under,heart_under)\n",
    "brain_muscle_under = intersector(brain_under,muscle_under)\n",
    "heart_muscle_under = intersector(muscle_under,heart_under)\n",
    "all_three_under = intersector(brain_heart_under,muscle_under)\n",
    "final_intersect = intersector(all_three_under, three_under)\n",
    "\n",
    "### CHANGE THE PATH to GenAge table\n",
    "GenAge = pd.read_csv('/Users/andrewstapran/Downloads/genage_models_export.tsv',sep='\\t')\n",
    "GenAge_genes_list = [element.upper() for element in list(GenAge['Gene Symbol'])]\n",
    "GenAge_genes_list_unique = list(set(GenAge_genes_list))\n",
    "\n",
    "#UNDER:\n",
    "GenAge_heart_under = intersector(heart_under,GenAge_genes_list_unique)\n",
    "GenAge_muscle_under = intersector(muscle_under,GenAge_genes_list_unique)\n",
    "GenAge_brain_under = intersector(brain_under,GenAge_genes_list_unique)\n",
    "\n",
    "#OVER:\n",
    "GenAge_heart_over = intersector(heart_over,GenAge_genes_list_unique)\n",
    "GenAge_muscle_over = intersector(muscle_over,GenAge_genes_list_unique)\n",
    "GenAge_brain_over = intersector(brain_over,GenAge_genes_list_unique)\n",
    "\n",
    "GenAge_total_over = intersector(three_over,GenAge_genes_list_unique)\n",
    "GenAge_total_under = intersector(three_under,GenAge_genes_list_unique)\n",
    "\n",
    "\n",
    "\n",
    "#function to build venn diagrams for intersection of three gene lists\n",
    "def plot_venn3(list1,list2,list3, name1, name2, name3):\n",
    "    venn3([set(list1), set(list2), set(list3)], \n",
    "          (name1, name2, name3))\n",
    "    plt.show()\n",
    "    \n",
    "\n",
    "    \n",
    "#function to build venn diagrams for intersection of three gene lists\n",
    "def plot_venn2(list1):\n",
    "    venn2([set(list1), set(GenAge_genes_list_unique)], \n",
    "          ('Three tissues overexpressed', 'GenAge'))\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "815a0bd9",
   "metadata": {},
   "source": [
    "<hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a94c3bc",
   "metadata": {},
   "source": [
    "### Conclusion\n",
    "With this code we were able to partially reproduce the results of original study. In particular:\n",
    "<ol>\n",
    "    <li>Immune function and inflammation are common GO features among gene up-regulated with age, especially in brain. This is consistent with inflammageing theory of ageing process</li>\n",
    "    <li>In muscle genes important for cell differentiation were updegulated. These cahnges can be associated withh deterioration of muscle tissue regeneration with age</li>\n",
    "    <li>Finally, metabolic genes were downregulated in heart muscle with age</li>\n",
    "</ol>\n",
    "Other results can be found in our <a href=\"https://github.com/d-kozhevnikova/Ageing-transcriptome-meta-analysis\">GitHub</a> repository.\n",
    "<br>\n",
    "<br>\n",
    "Credits for compiling the notebook and running the meta-regression analysis: <i>Andrei Stapran</i><br>\n",
    "<i>Moscow, March, 2021</i>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
